src/core/control/xojfile/XmlParser.cpp

```diff
@@ -0,0 +1,567 @@
+            XmlParserHelper::getAttribMandatory<fs::path>(xoj::xml_attrs::AUDIO_FILENAME_STR, attributeMap);
+    this->tempTimestamp = XmlParserHelper::getAttribMandatory<size_t>(xoj::xml_attrs::TIMESTAMP_STR, attributeMap);
+}
+
+void XmlParser::parseStrokeTag(const XmlParserHelper::AttributeMap& attributeMap) {
+    // tool
+    const auto tool =
+            XmlParserHelper::getAttribMandatory<StrokeTool>(xoj::xml_attrs::TOOL_STR, attributeMap, StrokeTool::PEN);
+    // color
+    const auto color = XmlParserHelper::getAttribColorMandatory(attributeMap, Colors::black);
+
+    // width
+    auto widthSV = XmlParserHelper::getAttribMandatory<std::string_view>(xoj::xml_attrs::WIDTH_STR, attributeMap, "1");
+    auto it = widthSV.begin();
+    auto end = widthSV.end();
+    double width{};
+    parseDouble(it, end, width);
+
+    // pressures
+    auto pressureSV = XmlParserHelper::getAttrib<std::string_view>(xoj::xml_attrs::PRESSURES_STR, attributeMap);
+    if (pressureSV) {
+        // MrWriter writes pressures in a separate field
+        it = pressureSV->begin();
+        end = pressureSV->end();
+    }
+    // Xournal and Xournal++ use the width field. `it` pointer is already in place.
+
+    double pressure{};
+    while (parseDouble(it, end, pressure)) {
+        this->pressureBuffer.emplace_back(pressure);
+    }
+
+    // fill
+    const auto fill = XmlParserHelper::getAttribMandatory<int>(xoj::xml_attrs::FILL_STR, attributeMap, -1, false);
+
+    // cap stype
+    const auto capStyle = XmlParserHelper::getAttribMandatory<StrokeCapStyle>(
+            xoj::xml_attrs::CAPSTYLE_STR, attributeMap, StrokeCapStyle::ROUND, false);
+
+    // line style
+    const auto lineStyle =
+            XmlParserHelper::getAttribMandatory<LineStyle>(xoj::xml_attrs::STYLE_STR, attributeMap, {}, false);
+
+    // audio filename and timestamp
+    const auto optFilename = XmlParserHelper::getAttrib<fs::path>(xoj::xml_attrs::AUDIO_FILENAME_STR, attributeMap);
+    if (optFilename && !optFilename->empty()) {
+        if (!this->tempFilename.empty()) {
+            g_warning("XML parser: Discarding audio timestamp element, because stroke tag contains \"fn\" attribute");
+        }
+        this->tempFilename = *optFilename;
+        this->tempTimestamp =
+                XmlParserHelper::getAttribMandatory<size_t>(xoj::xml_attrs::TIMESTAMP_STR, attributeMap, 0UL);
+    }
+
+    // forward data to handler
+    this->handler.addStroke(tool, color, width, fill, capStyle, lineStyle, std::move(this->tempFilename),
+                            this->tempTimestamp);
+
+    // Reset timestamp, filename was already moved from
+    this->tempTimestamp = 0;
+}
+
+void XmlParser::parseStrokeText(std::string_view text) {
+    std::vector<Point> pointVector;
+    pointVector.reserve(this->pressureBuffer.size());
+
+    auto it = text.begin();
+    const auto end = text.end();
+    double x{}, y{};
+    while (parseDouble(it, end, x) && parseDouble(it, end, y)) {
```

Comment:
It would be better if we'd get a warning in case the array has an odd number of values. Right now, the last value is simply discarded in that case, isn't it?